### PR TITLE
[AKS] Adding if_match and if_none_match headers for AKS RP 

### PR DESCRIPTION
**Pull request by InduSridhar**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] Have you run `azdev style <YOUR_EXT>` locally? (`pip install azdev` required)
- [ ] Have you run `python scripts/ci/test_index.py -q` locally? (`pip install wheel==0.30.0` required)
- [ ] My extension version conforms to the [Extension version schema](https://github.com/Azure/azure-cli/blob/release/doc/extensions/versioning_guidelines.md)

For new extensions:

- [ ] My extension description/summary conforms to the [Extension Summary Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/extensions/extension_summary_guidelines.md).

### About Extension Publish

There is a pipeline to automatically build, upload and publish extension wheels.  
Once your pull request is merged into main branch, a new pull request will be created to update `src/index.json` automatically.  
You only need to update the version information in file setup.py and historical information in file HISTORY.rst in your PR but do not modify `src/index.json`.

### CODE REVIEW

File: src/aks-preview/azext_aks_preview/_params.py [hunk 1]
```diff
@@ -1610,6 +1614,8 @@ def load_arguments(self, _):
             is_preview=True,
             help="delete an AKS nodepool by ignoring PodDisruptionBudget setting",
         )
+        c.argument("if_match")
+        c.argument("if_none_match")
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/_params.py:1618]: From API sepc, nodepool delete operation does not support request parameter `if_none_match`
https://github.com/Azure/azure-rest-api-specs/pull/28632

File: src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py [hunk 2]
```diff
@@ -4246,6 +4246,8 @@ def test_aks_create_with_windows(self, resource_group, resource_group_location):
                 "windows_admin_password": "replace-Password1234$",
                 "nodepool2_name": "npwin",
                 "ssh_key_value": self.generate_ssh_keys(),
+                "if_match": "*",
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:4249]: Queued [live test](https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=93139485&view=results) to validate the change
- test_aks_create_with_windows

**Review by FumingZhang — COMMENTED** [review]
lgtm

File: src/aks-preview/azext_aks_preview/_params.py (hunk 1 above)
Comment (InduSridhar) [reply src/aks-preview/azext_aks_preview/_params.py]: Removed in latest commit

**Comment by InduSridhar** [comment]
Successful run for the live test https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=93593533&view=results

File: src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py (hunk 2 above)
Comment (InduSridhar) [reply src/aks-preview/azext_aks_preview/tests/latest/test_aks_commands.py:4249]: Thanks, latest successful live test run here https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=93593533&view=results

**Comment by yonzhan** [comment]
Please fix conflicting files

**Comment by InduSridhar** [comment]
Latest live test run https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=93686897&view=results

File: src/aks-preview/azext_aks_preview/agentpool_decorator.py [hunk 3]
```diff
@@ -26,6 +26,7 @@
 from azure.cli.core.util import read_file_content
 from knack.log import get_logger
 from knack.prompting import prompt_y_n
+from azure.cli.core.util import sdk_no_wait
```
Comment (FumingZhang) [inline src/aks-preview/azext_aks_preview/agentpool_decorator.py:29]:
```suggestion
from azure.cli.core.util import (
    read_file_content,
    sdk_no_wait,
)
from knack.log import get_logger
from knack.prompting import prompt_y_n
```

Comment (InduSridhar) [reply src/aks-preview/azext_aks_preview/agentpool_decorator.py]: Updated

File: src/aks-preview/HISTORY.rst [hunk 4]
```diff
@@ -14,6 +14,7 @@ Pending
 
 4.0.0b4
 ++++++++
+* Add etag support to some aks commands for optimistic concurrency control.
```
Comment (matthchr) [inline src/aks-preview/HISTORY.rst:17]: minor, consider:
```suggestion
* Add etag support (--if-match, --if-none-match) to some aks commands for optimistic concurrency control.
```

File: src/aks-preview/azext_aks_preview/_help.py [hunk 5]
```diff
@@ -628,6 +628,12 @@
         - name: --bootstrap-container-registry-resource-id
           type: string
           short-summary: Configure container registry resource ID. Must use "Cache" as bootstrap artifact source.
+        - name: --if-match
+          type: string
+          short-summary: The ETag of the managed cluster. Omit this value to create a managed cluster.
```
Comment (matthchr) [inline src/aks-preview/azext_aks_preview/_help.py:633]: I don't love this description. We should talk about what happens if the etag doesn't match and what happens if it does.

On the second sentence, is: `This must not be specified when creating a new cluster` clearer than what you have?

File: src/aks-preview/azext_aks_preview/_params.py [hunk 6]
```diff
@@ -931,6 +931,8 @@ def load_arguments(self, _):
             is_preview=True,
             arg_type=get_enum_type(health_probe_modes),
         )
+        c.argument("if_match")
+        c.argument("if_none_match")
         # virtual machines
```
Comment (matthchr) [inline src/aks-preview/azext_aks_preview/_params.py:936]: Should we support these for `with self.argument_context("aks upgrade") as c:` too? I _think_ upgrade is just PUT MC?

Comment (matthchr) [reply src/aks-preview/azext_aks_preview/_params.py:936]: and/or PUT AP*

**Review by matthchr — APPROVED** [review]
This looks good to me for the most part, I left a few comments. Main person you should get to approve is Fuming.

File: src/aks-preview/azext_aks_preview/_help.py (hunk 5 above)
Comment (InduSridhar) [reply src/aks-preview/azext_aks_preview/_help.py]: Updated

File: src/aks-preview/azext_aks_preview/_params.py (hunk 6 above)
Comment (InduSridhar) [reply src/aks-preview/azext_aks_preview/_params.py:936]: Yes, you're right, thanks for catching that. Upgrade should be a PUT unless it's upgrading node image version in which case it is a POST and we do not support etags for POST.

File: src/aks-preview/HISTORY.rst (hunk 4 above)
Comment (InduSridhar) [reply src/aks-preview/HISTORY.rst]: Updated

**Comment by InduSridhar** [comment]
Successful test https://dev.azure.com/msazure/CloudNativeCompute/_build/results?buildId=94145423&view=results

File: src/aks-preview/HISTORY.rst [hunk 7]
```diff
@@ -14,6 +14,7 @@ Pending
 
 4.0.0b4
 ++++++++
+* Add etag support (--if-match, --if-none-match) to some aks commands for optimistic concurrency control.
```
Comment (FumingZhang) [inline src/aks-preview/HISTORY.rst:17]:
```suggestion
* Add etag support (--if-match, --if-none-match) to some aks commands for optimistic concurrency control.

4.0.0b4
++++++++
```

**Pull request merged**